### DISCUSSION TITLE
Does Trino support server-side cursors?

### DISCUSSION THREAD

**Discussion by rushilshah1** [trinodb/trino-python-client#395] [Q&A]
I am using the python client + sqlalchemy to query from trino. Does trino support server-side cursors? I tried adding `engine.connect().execution_options(stream_results=True)` but this was a no-op.

**Comment by mosabua** [comment]
From all I know cursors are not a construct with the SQL language and Trino does not support this. What are you actually trying to achieve?

**Reply by rushilshah1** [reply]
Trying to query trino using SQLAlchemy and stream the results using a server-side cursor to be more memory efficient (without loading the entire result set into memory)

**Reply by mosabua** [reply]
Trino is already taking care of being memory efficient and such. I would not worry about that on the client side or at least only optimize the client side.

**Reply by hashhar** [reply]
If you're using DB-API directly then you can use `fetchone` or `fetchmany` with a cursor size set explicitly to avoid pulling all results into memory - note that you'd need to do that in a loop to ensure all results are retreived. `fetchall` on the other hand would pull all rows and block until all rows are retreived.

SQLA also has similar methods - see https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CursorResult.fetchmany
